### PR TITLE
Implement multi-block uncover-live

### LINKED ISSUE #28: Uncover live
**TODO**
- [x] uncover-live should handle bools `#t` `#f`
  * `edit:` not needed because `select-instructions` converts bools to 1 or 0
- [ ] uncover-live should handle handle new x86 constructs (see below)
  * `edit:` perhaps no changes/few changes needed here because uncover-live helpers are written generally and perviously the special cases were `addq` and `negq` and it doesn't seem like any new special cases were introduced in R2
- [x] uncover-live should work with multiple blocks (not just 1)
  - [x] calculate the live-after-sets for all blocks/branches on Control-Flow-Graph
    - [x] **_no successor block_** means **_empty live-after-set_**
    - [x] **_∃ successor block_** means **_calculate live-after-sets for successors first_**
    - [x] If you reach a branch then _union the live-after-sets of all of the successor blocks_

**Example Input**
```racket
> p
'(program
  ()
  ((block176 (return x))
   (start (if (eq? #t #t) (goto block177) (goto block178)))
   (block178 (seq (assign x 3) (goto block176)))
   (block177 (seq (assign x 2) (goto block176)))))

> (select-instructions p)
'(program
  ()
  ((label block176)
   (movq (var x) (reg rax))
   (jmp conclusion)
   (label start)
   (cmpq (int 1) (int 1))
   (jmp-if e block177)
   (jmp block178)
   (label block178)
   (movq (int 3) (var x))
   (jmp block176)
   (label block177)
   (movq (int 2) (var x))
   (jmp block176)))
>
```
* consider the `start` block. 
  * It has a successor `block178`
    * which has a successor `block176`
      * which has **no successors**... let's call it a "**_basic block_**"
        * which has `life-after-set = {}`
* Fun facts:
  * the input program is a _Control-Flow-Graph_ and is a _Directed-Acyclic-Graph_

**Algorithm**
* calculate the live-after-sets in _**reverse topological order**_ starting from the "basic blocks"
  * `tsort` and `transpose` functions exist in [`graph`](https://docs.racket-lang.org/graph/index.html) and could be useful for this purpose.

image

### PR DESCRIPTION
**Pull request by mfekadu**
resolves #28

### CODE REVIEW

File: testing/uncover-live-tests.rkt [hunk 1]
```diff
@@ -164,12 +167,61 @@
...
+
+(define expect-blocks
+  `((block
+     ; init set of block176 is (set 'x)
+     ; Live-After-Set of the "label" instruction is (set 'x)
+     ; after that nothing is alive because it will retq
+     (,(set 'x),(set 'x) ,(set)                   ,(set))
+     (label block176)     (movq (var x) (reg rax)) (jmp conclusion))
+    (block
+     (,(set),(set)    ,(set)                 ,(set))
```
Comment (johnvtan) [inline testing/uncover-live-tests.rkt]: Shouldn't `x` be alive on the `jmp block176` instruction?

File: testing/uncover-live-tests.rkt [hunk 2]
```diff
@@ -164,12 +167,61 @@
...
+     (label block178)  (movq (int 3) (var x)) (jmp block176))
+    (block
+     (,(set),(set)    ,(set)                 ,(set))
```
Comment (johnvtan) [inline testing/uncover-live-tests.rkt]: Same as other comment

File: testing/uncover-live-tests.rkt [hunk 3]
```diff
@@ -164,12 +167,61 @@
...
+
+(check-equal? expect-blocks (get-LAS-from-blocks given-hash given-g))
+
 ; ==================================================
 ; TEST uncover-live
 ; ==================================================
 
-(define given1-uncover `(program () (start (block () ,given1-glas))))
-(define expect1-uncover `(program () (start (block ,expect1-glas ,given1-glas))))
+; NOTE: the ,@ syntax will splice the elements of list into a "quasiquoted list"
```
Comment (johnvtan) [inline testing/uncover-live-tests.rkt:217]: cool!

**Review by johnvtan — CHANGES_REQUESTED** [review]
I think the test cases might be wrong - maybe I'm misunderstanding something though

File: testing/uncover-live-tests.rkt (hunk 1 above)
Comment (johnvtan) [reply testing/uncover-live-tests.rkt]: I realize now that this wasn't the best line to make a reference to

**Comment by mfekadu** [comment]
## Setup Before Iterating Over Blocks
**Consider the following:**
* `given-blocks` 
* the mutable `given-hash` 
* and `given-g`
```racket
(require graph)
(require "src/uncover-live.rkt")

(define given-blocks
  '((block () (label block177) (movq (int 2) (var x)) (jmp block176))
    (block () (label block178) (movq (int 3) (var x)) (jmp block176))
    (block () (label start) (cmpq (int 1) (int 1)) (jmp-if e block177) (jmp block178))
    (block () (label block176) (movq (var x) (reg rax)) (jmp conclusion))))

; NOTE: this hashtable is mutable
(define given-hash (make-hash
                    `((block177 ,@(first given-blocks))
                      (block178 ,@(second given-blocks))
                      (start ,@(third given-blocks))
                      (block176 ,@(fourth given-blocks)))))

(define dag_adj_matrix_of_given_blocks
  '((block177 block176)
    (block178 block176)
    (start block177 block178)
    (block176 conclusion)))
(define given-g (unweighted-graph/adj dag_adj_matrix_of_given_blocks))
```

**The `get-LAS-from-blocks` function will iterate over the vertices `vs`**
```racket
> (define vs (remove 'conclusion (reverse (tsort given-g))))
> vs
'(block176 block178 block177 start)
```

## First Iteration
**On the _first_ iteration consider the block `b`**
```racket
> (define b (hash-ref given-hash (first vs)))
> b
'(block () (label block176) (movq (var x) (reg rax)) (jmp conclusion))
```

**The _initial set_ for calculating the LAS here is the _empty set_**
```racket
> (define init_set (set))
> (define instrs (drop b 2))
> instrs
'((label block176) (movq (var x) (reg rax)) (jmp conclusion))
> (get-live-after-sets instrs init_set)
(list (set 'x) (set 'x) (set) (set))
```

## Second Iteration
**On the _second_ iteration consider the block `b`**
```racket
> (define b (hash-ref given-hash (second vs)))
> b
'(block () (label block178) (movq (int 3) (var x)) (jmp block176))
```

**The _initial set_ for calculating the LAS here is the _previously calculated set_**
```racket
> (define prev-LAS-list (list (set 'x) (set 'x) (set) (set)))
> (define init_set  (first prev-LAS-list) )
> init_set
(set 'x)
> (define instrs (drop b 2))
> instrs
'((label block178) (movq (int 3) (var x)) (jmp block176))
```

**Recall that the calculation of the LAS is:**
* `{ L_after - Write_vars } U { Read_vars }`
```racket
> (get-live-after-sets instrs init_set)
(list (set) (set) (set) (set))
```

## Third Iteration
**On the _third_ iteration consider the block `b`**
```racket
> (define b (hash-ref given-hash (third vs)))
> b
'(block () (label block177) (movq (int 2) (var x)) (jmp block176))
```

**The _initial set_ for calculating the LAS here is the _previously calculated set_**
```racket
> (define prev-LAS-list (list (set) (set) (set) (set)) )
> (define init_set  (first prev-LAS-list) )
> init_set
(set)
> (define instrs (drop b 2))
> instrs
'((label block177) (movq (int 2) (var x)) (jmp block176))
> (get-live-after-sets instrs init_set)
(list (set) (set) (set) (set))
```

## Fourth Iteration (last)
**On the _fourth_ iteration consider the block `b`**
```racket
> (define b (hash-ref given-hash (fourth vs)))
> b
'(block
  ()
  (label start)
  (cmpq (int 1) (int 1))
  (jmp-if e block177)
  (jmp block178))
```

**The _initial set_ for calculating the LAS here is the _previously calculated set_**
```racket
> (define prev-LAS-list (list (set) (set) (set) (set)) )
> (define init_set  (first prev-LAS-list) )
> init_set
(set)
> (define instrs (drop b 2))
> instrs
'((label start) (cmpq (int 1) (int 1)) (jmp-if e block177) (jmp block178))
> (get-live-after-sets instrs init_set)
(list (set) (set) (set) (set) (set))
```

## A list of each LAS from the above iterations
```
(list 
    (list (set 'x) (set 'x) (set) (set))
    (list (set) (set) (set) (set))
    (list (set) (set) (set) (set))
    (list (set) (set) (set) (set) (set)))
```


## The list of live-after sets above matches the expectation from the test cases copied below:
```racket
(define expect-blocks
  `((block
     ; init set of block176 is (set 'x)
     ; Live-After-Set of the "label" instruction is (set 'x)
     ; after that nothing is alive because it will retq
     (,(set 'x),(set 'x) ,(set)                   ,(set))
     (label block176)     (movq (var x) (reg rax)) (jmp conclusion))
    (block
     (,(set),(set)    ,(set)                 ,(set))
     (label block178)  (movq (int 3) (var x)) (jmp block176))
    (block
     (,(set),(set)    ,(set)                 ,(set))
     (label block177)  (movq (int 2) (var x)) (jmp block176))
    (block
     (,(set),(set) ,(set)                  ,(set)               ,(set))
     (label start)  (cmpq (int 1) (int 1))  (jmp-if e block177)  (jmp block178))))

(check-equal? expect-blocks (get-LAS-from-blocks given-hash given-g))
```


## But wait let's inspect the blocks with the intuition that "_a variable is alive after it is written to and until the final read from it_"

**Recall that the calculation of the LAS is:**
* `{ L_after - Write_vars } U { Read_vars }`

**Here are the `given-blocks` from above but ordered in topological-order according to the control-flow-graph**
```racket
'( (block ()                               
    (label start)                           ; {   }
    (cmpq (int 1) (int 1))                  ; {   }
    (jmp-if e block177) (jmp block178))     ; {   }
(block ()                                        
    (label block177)                        ; {   }
    (movq (int 2) (var x))                  ; { x } Bugs! Bugs! Bugs! Woohoo!
    (jmp block176))                         ; { x } And here!
(block ()                                        
    (label block178)                        ; {   }
    (movq (int 3) (var x))                  ; { x } Here too!
    (jmp block176))                          ; { x }    Here's the bug! 'x should be here, but isn't in iterations above
(block ()                                        
    (label block176)                        ; { x }
    (movq (var x) (reg rax))                ; { x }
    (jmp conclusion))) )                    ; {   }
```

## So, yes there does seem to be a mistake on the set for 
## `(jmp block176))`

**Comment by mfekadu** [comment]
@johnvtan, sorry for the giant comment, but does the last section titled "But wait" show the issue you mentioned in your comment?

**Review by mfekadu — COMMENTED** [review]
fixed issues with tests in uncover-live and new tests are now passing

**Comment by mfekadu** [comment]
```
> ./run_tests.sh

Running assign-homes tests...
assign homes tests finished

Running ec-tests...
ec tests finished

Running patch instructions tests...
patch instructions tests pass

Running print-x86 tests...
print-x86 tests finished

Running rco tests...
rco tests finished

Running select instructions tests...
select instructions tests finished

Running uncover locals tests...
uncover locals tests finished

Running uncover-live tests...
uncover-live tests finished

Running uniquify tests...
uniquify tests finished

Running graph tests...
Graph tests finished running

Running make-interference tests
make interference tests done

Running allocate-registers tests
allocate registers test done

Running typecheck tests
Typecheck tests finished

Running shrink tests
shrink tests finished
>
```

**Review by johnvtan — APPROVED** [review]
Looks good!

**Pull request merged**